src/contract-access-controller.js

```diff
@@ -2,43 +2,57 @@
 
 const AccessController = require('./access-controller-interface')
 
-const type = 'eth-contract/my-contract-ac'
+const type = 'eth-contract'
 
 class ContractAccessController extends AccessController {
   constructor(web3, abi, address) {
     super()
     this.web3 = web3
     this.abi = abi
-    // this.contract = new this.web3.eth.Contract(abi, address)
     this.contractAddress = address
   }
 
   // Returns the type of the access controller
   static get type () { return type }
 
-  async load (address, options) {}
+  async load () {
+    this.contract = new this.web3.eth.Contract(this.abi, this.contractAddress)
+  }
 
   async save () {
-    return { 
-      contractAddress: this.contractAddress, 
-      abi: this.abi 
+    return {
+      contractAddress: this.contractAddress,
+      abi: this.abi
     }
   }
 
+  static isValidEthAddress (web3, address) {
+    return web3.utils.isAddress(address)
+  }
+
   async canAppend (entry, identityProvider) {
     // Write the custom access control logic here
-    // return await this.contract.methods.haveTheyPaidTheDeposit(entry.identity.id).call()
-    return await this.contract.methods.isPermitted(identifier, this.web3.utils.fromAscii('write')).call()
+    if (!ContractAccessController.isValidEthAddress(this.web3, entry.identity.id)) {
+      console.warn(`WARNING: "${entry.identity.id}" is not a valid eth address`)
+      return Promise.resolve(false)
+    }
+    return await this.contract.methods.isPermitted(entry.identity.id, this.web3.utils.fromAscii('write')).call()
   }
 
-  async grant (identifier, capability) {
-    // return await this.contract.methods.paidTheirDeposit(identifier).call()
-    return await this.contract.methods.grantCapability(identifier, this.web3.utils.fromAscii(capability)).send( { from: this.primaryAccount } )
+  async grant (capability, identifier, options = {}) {
+    if (!ContractAccessController.isValidEthAddress(this.web3, identifier)) {
+      console.warn(`WARNING: "${identifier}" is not a valid eth address`)
+      return Promise.resolve(false)
+    }
+    return await this.contract.methods.grantCapability(identifier, this.web3.utils.fromAscii(capability)).send(options)
```

Comment:
It would make sense perhaps to bring in the `primaryAccount` option already in the factory/constructor so that it can be called here as `this.primaryAccount`. That way the user doesn't have to add the `{ from: accounts[0] }` to every call to `grant()` but can pass the `primaryAccount` in only once.

Reply:
:+1: agree but adding it as an optional argument as some peers may just want to ask the state of the contract without changing it, which doesn't require an account

Reply:
Makes sense 👍